### PR TITLE
Add alternative names for the server binaries to hyperkube

### PR DESCRIPTION
**Pull request by luxas**
**What this PR does / why we need it**:

Right now one can't swap a server image to the hyperkube image without touching the `command` field in the yaml spec, and that's daunting and leading to extra and unnecessary logic for example in kubeadm.

This makes the hyperkube image directly swappable, so now `/usr/local/bin/kube-*` is a portable first argument (or simply `kube-*` if there's a shell).

**Special notes for your reviewer**:

**Release note**:

```release-note
Align the hyperkube image to support running binaries at /usr/local/bin/ like the other server images
```
@jessfraz @thockin @ixdy

### CODE REVIEW

File: cluster/images/hyperkube/Dockerfile [hunk 1]
```diff
@@ -76,8 +78,11 @@ RUN ln -s /hyperkube /apiserver \
  && ln -s /hyperkube /kubectl \
  && ln -s /hyperkube /kubelet \
```
Comment (errordeveloper) [inline cluster/images/hyperkube/Dockerfile]: while at it, may be add symlinks in `/usr/local/bin` for these two also? It's kind of annoying these aren't in the path...

File: cmd/hyperkube/kube-apiserver.go [hunk 2]
```diff
@@ -27,8 +27,10 @@ func NewKubeAPIServer() *Server {
 	s := options.NewServerRunOptions()
 
 	hks := Server{
-		SimpleUsage: "apiserver",
-		Long:        "The main API entrypoint and interface to the storage system.  The API server is also the focal point for all authorization decisions.",
+		name:            "apiserver",
```
Comment (ixdy) [inline cmd/hyperkube/kube-apiserver.go]: is it necessary to add `name` for all of these? doesn't `Name()` DTRT already?

File: cmd/hyperkube/hyperkube.go [hunk 3]
```diff
@@ -55,7 +55,7 @@ func (hk *HyperKube) AddServer(s *Server) {
 // FindServer will find a specific server named name.
 func (hk *HyperKube) FindServer(name string) (*Server, error) {
 	for _, s := range hk.servers {
-		if s.Name() == name {
+		if s.Name() == name || s.AlternativeName == name {
```
Comment (ixdy) [inline cmd/hyperkube/hyperkube.go]: it's not 100% correct, but I wonder if just additionally checking for a `kube-` prefix would be an easier way of accomplishing the same thing.

File: cmd/hyperkube/kube-apiserver.go (hunk 2 above)
Comment (luxas) [inline cmd/hyperkube/kube-apiserver.go]: It's not 100% required, but I think it's a little bit cleaner than just relying on SimpleUsage for this

File: cmd/hyperkube/hyperkube.go (hunk 3 above)
Comment (luxas) [inline cmd/hyperkube/hyperkube.go]: Yes, I thought about that as well, but I think the code changes are so small in any case it's worth to do it "the real way"

**Comment by luxas** [comment]
@ixdy Does this look good to you if I fix the nit @errordeveloper mentioned?

**Comment by luxas** [comment]
@k8s-bot gci gke e2e test this
@k8s-bot gci gce e2e test this

File: cmd/hyperkube/kube-apiserver.go (hunk 2 above)
Comment (ixdy) [inline cmd/hyperkube/kube-apiserver.go]: not critical, but if we're going to add `name`s for everything, maybe also add them for the federation- servers and then just simplify `Name()`?

File: cmd/hyperkube/server.go [hunk 4]
```diff
@@ -30,9 +30,10 @@ type serverRunFunc func(s *Server, args []string) error
 
 // Server describes a server that this binary can morph into.
 type Server struct {
-	SimpleUsage string        // One line description of the server.
-	Long        string        // Longer free form description of the server
-	Run         serverRunFunc // Run the server.  This is not expected to return.
+	SimpleUsage     string        // One line description of the server.
+	Long            string        // Longer free form description of the server
+	Run             serverRunFunc // Run the server.  This is not expected to return.
+	AlternativeName string
```
Comment (ixdy) [inline cmd/hyperkube/server.go]: maybe add a comment for this?

**Comment by ixdy** [comment]
/lgtm
with a few minor comments/suggestions

**Comment by mikedanese** [comment]
/approve

please squash

**Comment by mikedanese** [comment]
Also @ixdy has a comment.

**Comment by luxas** [comment]
The commits are logical, one for modifying `cmd/hyperkube`, one for modifying the image.

May I please add the comment in a future PR if that's important? I'd like to get this in during the weekend in time for v1.6.0-alpha.2 (so kubeadm can be simplified) and I'm going to sleep now :)

**Comment by ixdy** [comment]
I'm fine with addressing my two comments in a separate PR.

**Pull request merged**